### ISSUE TITLE
Pure MM on ProBit - order price quantum seems to be 0

### ISSUE THREAD

**Issue by tab00** [CoinAlpha/hummingbot#4312]
**Describe the bug**
When order_optimization_enabled is true the new sell limit order should have price = best ask - order price quantum as described in https://docs.hummingbot.io/market-making/order-optimization. However the price of the order is exactly best ask, i.e. as if quantum is 0.

**Steps To Reproduce**
Here are the steps to reproduce the issue (see attachments in the section below):
1. Set up your configuration file with these parameters: ...
ask_spread: 0.001
order_optimization_enabled: true
ask_order_optimization_depth: 0

**Release version**
0.43.1

**Attachments**
[logs_conf_pure_mm_probit.log](https://github.com/CoinAlpha/hummingbot/files/7246043/logs_conf_pure_mm_probit.log)
[conf_pure_mm_probit.yml.txt](https://github.com/CoinAlpha/hummingbot/files/7246044/conf_pure_mm_probit.yml.txt)

**Comment by Jurelll** [comment]
Hi @tab00, `order_optimization_enabled` Allows your bid and ask order prices to be adjusted based on the current top bid and ask prices in the market, and you can set `ask_order_optimization_depth` and `bid_order_optimization_depth` as they work hand in hand. I will try and test your configurations as well.

You may check these links below:
https://docs.hummingbot.io/market-making/order-optimization/#ask_order_optimization_depth
https://docs.hummingbot.io/market-making/order-optimization/#bid_order_optimization_depth

**Comment by tab00** [comment]
> your `ask_order_optimization_depth` and `bid_order_optimization_depth` is set to 0
... as they are in https://docs.hummingbot.io/market-making/order-optimization
That is not relevant to the problem. I've just tried it anyway with a non-0 value and the problem remains.

**Comment by tab00** [comment]
![image](https://user-images.githubusercontent.com/12715448/135205995-fecbc370-2a1d-454b-82c0-84cf0dbbf574.png)

`11:50:30 - pure_market_making - (PROB-USDT) Creating 1 ask orders at (Size, Price): ['2E+2 PROB, 0.308 USDT']`

Price should be 0.307 USDT.

**Comment by gerald-hb** [comment]
Ask price is = 0.308 x 1+00001 (your spread converted to decimals) = 0.30800308 
Your ask price is is almost same with the current best ask price, since the orderbook only show up to 3 decimal place only we can't really tell if you get the best ask already.

**Comment by tab00** [comment]
> Ask price is = 0.308 x 1+00001 (your spread converted to decimals) = 0.30800308

That's not correct as mid price is used for adding the spread, not ask price.
Since `order_optimization_enabled` is true, if the resulting price is within the market bid-ask spread, price should then be adjusted to be ask - quantum. quantum should be 0.001 in this case, not 0.

**Comment by tab00** [comment]
See line https://github.com/CoinAlpha/hummingbot/blob/9bf8e0cc16c00e7458185a2a772a46f1b08f56db/hummingbot/strategy/pure_market_making/pure_market_making.pyx#L922

**Comment by tab00** [comment]
@dennisocana why change it to support and not bug? Order price quantum should never be 0.

**Comment by dennisocana** [comment]
@tab00 once QA or engineering has validated, I will change it back to bug.

**Comment by dennisocana** [comment]
> ![image](https://user-images.githubusercontent.com/12715448/135205995-fecbc370-2a1d-454b-82c0-84cf0dbbf574.png)
> 
> `11:50:30 - pure_market_making - (PROB-USDT) Creating 1 ask orders at (Size, Price): ['2E+2 PROB, 0.308 USDT']`
> 
> Price should be 0.307 USDT.

Why do you think the price should be 0.307 USDT? Order optimization doesn't adjust your order closer to the mid price. Instead, it should adjust your spreads farther away.

**Comment by tab00** [comment]
> Order optimization doesn't adjust your order closer to the mid price

Yes it does, and that's the behavior I want as a trader - I want my order to be the best in the market.

See the example in https://docs.hummingbot.io/market-making/order-optimization/#sample-configuration

**Comment by tab00** [comment]
Also see this line that creates price just above bid: https://github.com/CoinAlpha/hummingbot/blob/9bf8e0cc16c00e7458185a2a772a46f1b08f56db/hummingbot/strategy/pure_market_making/pure_market_making.pyx#L927

And this line that creates price just below ask: https://github.com/CoinAlpha/hummingbot/blob/9bf8e0cc16c00e7458185a2a772a46f1b08f56db/hummingbot/strategy/pure_market_making/pure_market_making.pyx#L945

**Comment by tab00** [comment]
The "optimization" is about not having the price any better than it needs to be, hence the use of the price quantum.

**Comment by dennisocana** [comment]
In that case, that's not how the current order optimization works and should be treated as a new feature. The objective of order optimization as described in the example shows pushing your spreads farther away from the mid price just beside the next best price in the orderbook. For example:

Best ask = 100
Best bid = 90
Mid price = 95

If your spreads are too tight that it puts your initial sell order at 95 (as an example), it will push it back and adjust to 99. It doesn't make you "skip" the line by putting your orders on top if you're initial spreads are not there.

**Comment by tab00** [comment]
> If your spreads are too tight that it puts your initial sell order at 95 (as an example), it will push it back and adjust to 99

99 would actually be correct and what I want. Order price quantum would be 1 in this example.

The problem right now is that quantum is 0. Using your example, the order would be at 100 instead of 99.

**Comment by tab00** [comment]
> should be treated as a new feature

Not a new feature. I just want it to work as documented.

**Comment by dennisocana** [comment]
> > If your spreads are too tight that it puts your initial sell order at 95 (as an example), it will push it back and adjust to 99
> 
> 99 would actually be correct and what I want. Order price quantum would be 1 in this example.
> 
> The problem right now is that quantum is 0. Using your example, the order would be at 100 instead of 99 because of this bug.

For @CoinAlpha/qa-team validation. Please see user's comment above and confirm.

**Comment by rapcmia** [comment]
Hi @tab00, We tried to look into it using a different connectors/markets and found out that one factor is how the exchange display's the market prices on the orderbook. On this reported ticket, I've tested Probit and use DAI-USDT as an example. Set both bid/ask spreads to `0.00001` (closest I can think) with order_amount of 11 and it seems to work just fine. It created an order above the best bid and below the best ask.

![order optmization 3](https://user-images.githubusercontent.com/73840223/135835314-e5f6291b-bbc8-42fb-a2ec-cc62743c0b59.png)

I was able to replicate the issue on a different exchange but with same pairs as Probit (USDT/DAI). On the screenshow below, you will noticed that the orders should've been the same as the example above but its not. So I tried to calculate how the orders are created using the same config

![order optmization 4](https://user-images.githubusercontent.com/73840223/135835755-dd053fd9-2946-4590-bc61-b0c3ca5840ea.PNG)

For the best bid ask : mid_price * (1+ask_spread)
So we are expecting the bot should create ask order of `0.9999599995`, however since the order book can only display upto 4 decimal places for the prices, the bot created the order to `1.0000` instead resulting my order to be added to the current best ask.
